consensus/istanbul/backend/announce.go

```diff
@@ -1006,60 +1098,79 @@ func (sb *Backend) upsertAndGossipVersionCertificateEntries(entries []*vet.Versi
 	return nil
 }
 
-// updateAnnounceVersion will synchronously update the announce version.
-// Must be called in a separate goroutine from the announceThread to avoid
-// a deadlock.
-func (sb *Backend) updateAnnounceVersion() {
+// UpdateAnnounceVersion will asynchronously update the announce version.
+func (sb *Backend) UpdateAnnounceVersion() {
 	sb.updateAnnounceVersionCh <- struct{}{}
-	<-sb.updateAnnounceVersionCompleteCh
+}
+
+// GetAnnounceVersion will retrieve the current announce version.
+func (sb *Backend) GetAnnounceVersion() uint {
+	sb.announceVersionMu.RLock()
+	defer sb.announceVersionMu.RUnlock()
+	return sb.announceVersion
 }
 
 // setAndShareUpdatedAnnounceVersion generates announce data structures and
 // and shares them with relevant nodes.
 // It will:
 //  1) Generate a new enode certificate
-//  2) Send the new enode certificate to this node's proxy if one exists
-//  3) Send the new enode certificate to all peers in the validator conn set
-//  4) Generate a new version certificate
-//  5) Gossip the new version certificate to all peers
+//  2) Multicast the new enode certificate to all peers in the validator conn set (note that if this is a proxied validator, it's multicast
+//     message will be wrapped within a forward message to the proxy (which will in turn send the enode certificate to remote validators).
+//  3) Generate a new version certificate
+//  4) Gossip the new version certificate to all peers
 func (sb *Backend) setAndShareUpdatedAnnounceVersion(version uint) error {
 	logger := sb.logger.New("func", "setAndShareUpdatedAnnounceVersion")
 	// Send new versioned enode msg to all other registered or elected validators
 	validatorConnSet, err := sb.retrieveValidatorConnSet()
 	if err != nil {
 		return err
 	}
-	enodeCertificateMsg, err := sb.generateEnodeCertificateMsg(version)
-	if err != nil {
-		return err
-	}
-	sb.setEnodeCertificateMsg(enodeCertificateMsg)
-	// Send the new versioned enode msg to the proxy peer
-	if sb.config.Proxied && sb.proxyNode != nil && sb.proxyNode.peer != nil {
-		err := sb.sendEnodeCertificateMsg(sb.proxyNode.peer, enodeCertificateMsg)
-		if err != nil {
-			logger.Error("Error in sending versioned enode msg to proxy", "err", err)
-			return err
-		}
-	}
+
 	// Don't send any of the following messages if this node is not in the validator conn set
 	if !validatorConnSet[sb.Address()] {
 		logger.Trace("Not in the validator conn set, not updating announce version")
 		return nil
 	}
-	payload, err := enodeCertificateMsg.Payload()
+
+	enodeCertificateMsgs, err := sb.generateEnodeCertificateMsgs(version)
 	if err != nil {
 		return err
 	}
-	destAddresses := make([]common.Address, len(validatorConnSet))
-	i := 0
+
+	if len(enodeCertificateMsgs) > 0 {
```

Comment:
Is there an edge case here where if the validator has no external enodes at a given time (e.g. all proxies are removed) it will not set the enode certificate map to the correct state, which is empty? If I read this correctly, then we should cover that edge case by setting the map to empty and logging a warning or error about this issue.

Reply:
It will create enode certificates for all proxies that are added to the validator (and not necessarily connected).  I actually don't really see much downside in doing this.

Once a proxy does connect to the validator, then new enode certificates will be created, since the announce version will be updated.